File: app/src/main/java/com/rossyn/blocktiles/game2048/data/prefs/SharedPref.java
package com.rossyn.blocktiles.game2048.data.prefs;

import android.content.Context;
import android.content.SharedPreferences;

public class SharedPref {

    public static final String SHARED_PREF_FILE_NAME = "2048";

    public static final String MUTE_MUSIC = "mute_music";
    public static final String MUTE_SOUNDS = "mute_sounds";
    public static final String TUTORIAL_PLAYED = "tutorial_played";
    public static final String TOP_SCORE = "TopScore";

    private final SharedPreferences preferences;

    public SharedPref(Context context) {
        preferences = context.getApplicationContext().getSharedPreferences(SHARED_PREF_FILE_NAME, Context.MODE_PRIVATE);
    }

    public void setBoolean(String key, boolean value) {
        preferences.edit().putBoolean(key, value).apply();
    }

    public boolean getBoolean(String key) {
        return preferences.getBoolean(key, false);
    }

    public void setString(String key, String value) {
        preferences.edit().putString(key, value).apply();
    }

    public String getString(String key) {
        return preferences.getString(key, "");
    }

    public void setInt(String key, int value) {
        preferences.edit().putInt(key, value).apply();
    }

    public int getInt(String key) {
        return preferences.getInt(key, 0);
    }

    public void setLong(String key, long value) {
        preferences.edit().putLong(key, value).apply();
    }

    public long getLong(String key) {
        return preferences.getLong(key, 0L);
    }

    /**
     * Removes the preference value for the given key.
     *
     * @param key The key whose value needs to be removed.
     */
    public void remove(String key) {
        preferences.edit().remove(key).apply();
    }

    /**
     * Clears all preferences.
     */
    public void clear() {
        preferences.edit().clear().apply();
    }
}

File: app-kts/src/main/java/com/rossyn/blocktiles/game2048/data/prefs/SharedPref.kt
package com.rossyn.blocktiles.game2048.data.prefs

import android.content.Context
import android.content.SharedPreferences
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.properties.ReadWriteProperty
import kotlin.reflect.KProperty

/**
 * This object handles the SharedPreferences operations.
 * It provides methods to save, retrieve, and delete preferences.
 * This class demonstrates the use of SharedPreferences for storing user preferences.
 *
 */
@Singleton
class SharedPref @Inject constructor(
    private val preferences: SharedPreferences, @ApplicationContext private val context: Context
) {
    companion object {
        const val MUTE_MUSIC = "mute_music"
        const val MUTE_SOUND = "mute_sounds"
        const val TUTORIAL_PLAYED = "tutorial_played"
        const val TOP_SCORE = "Top_score"
    }

    /**
     * Registers a listener to be notified when a shared preference is changed.
     *
     * @param listener The callback that will run when a shared preference is changed.
     */
    fun registerOnSharedPreferenceChangeListener(listener: SharedPreferences.OnSharedPreferenceChangeListener) {
        Timber.tag("SharedPref").d("registerOnSharedPref")
        preferences.registerOnSharedPreferenceChangeListener(listener)
    }

    /**
     * Unregisters a previously registered listener.
     *
     * @param listener The callback to be unregistered.
     */
    fun unregisterOnSharedPreferenceChangeListener(listener: SharedPreferences.OnSharedPreferenceChangeListener) {
        Timber.tag("SharedPref").d("unregisterOnSharedPref")
        preferences.unregisterOnSharedPreferenceChangeListener(listener)
    }

    var isMuteMusic by preference(MUTE_MUSIC, false)
    var isMuteSound by preference(MUTE_SOUND, false)
    var isTutorialPlayed by preference(TUTORIAL_PLAYED, false)

    /**
     * Generic preference delegate for all types (nullable and non-nullable).
     * @param name The key for the preference.
     * @param defaultValue The default value for the preference (nullable).
     */
    private inline fun <reified T> preference(name: String, defaultValue: T) =
        object : ReadWriteProperty<SharedPref, T> {

            override fun getValue(thisRef: SharedPref, property: KProperty<*>): T {
                return when (T::class) {
                    Boolean::class -> preferences.getBoolean(name, defaultValue as Boolean) as T
                    Int::class -> preferences.getInt(name, defaultValue as Int) as T
                    Long::class -> preferences.getLong(name, defaultValue as Long) as T
                    Float::class -> preferences.getFloat(name, defaultValue as Float) as T
                    String::class -> preferences.getString(name, defaultValue as String?) as T
                    else -> {
                        Timber.tag("SharedPref").d("Type not supported: ${T::class.java}")
                        defaultValue // Return default value if type is unsupported
                    }
                }
            }

            override fun setValue(thisRef: SharedPref, property: KProperty<*>, value: T) {
                CoroutineScope(Dispatchers.IO).launch {
                    preferences.edit().apply {
                        when (value) {
                            is Boolean -> putBoolean(name, value)
                            is Int -> putInt(name, value)
                            is Long -> putLong(name, value)
                            is Float -> putFloat(name, value)
                            is String? -> putString(name, value)
                            else -> {
                                Timber.tag("SharedPref").d("Type not supported: ${T::class.java}")
                                // Do nothing if type is unsupported
                            }
                        }
                    }.apply()
                }
            }
        }
}